### PR TITLE
Add microsoft distribution of the JDK.

### PR DESCRIPTION
**Pull request by brendandburns**
**Description:**
Initial add of the Microsoft OpenJDK distribution

**Related issue:**
N/A

**Check list:**
- [X] Mark if documentation changes are required.
- [X] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Comment by brendandburns** [comment]
This was based off of #225 thanks for making it easy!

**Comment by brendandburns** [comment]
cc @brunoborges

File: README.md
Comment (brunoborges) [inline README.md:61]: Thanks @brendanburns! One quick suggestion here on the name: `Microsoft Build of OpenJDK`

**Comment by brunoborges** [comment]
@ethomson this is a great PR.

**Comment by ethomson** [comment]
Thanks @brendandburns!  We'll take a look.  👀

File: README.md
Comment (brendandburns) [reply README.md]: Done.

**Comment by brendandburns** [comment]
Comment addressed and I fixed formatting to follow prettier style guidelines.

**Comment by brendandburns** [comment]
Hold on this for a second, I realized that the Windows URLs aren't right, I need to substitute ".zip" for "tar.gz" I will get to that sometime today.

**Comment by brendandburns** [comment]
Ok, this is now fixed and I believe should be ready for final review.

**Review by dikehtaw — APPROVED** [review]

File: README.md
Comment (dikehtaw) [inline README.md]: brendandburns:main

**Review by dikehtaw — COMMENTED** [review]
gh pr checkout 252

**Comment by dmitry-shibanov** [comment]
Hello @brendanburns. Thank you for you pull request. Will you add logic to automatically grab all available versions ?

File: README.md
Comment (brcrista) [inline README.md:61]:
```suggestion
| `microsoft` | Microsoft Build of OpenJDK | [Link](https://www.microsoft.com/openjdk) | [Link](https://docs.microsoft.com/java/openjdk/faq)
```

I believe `docs.microsoft.com` will redirect to the appropriate page for the user's language setting

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [inline src/distributions/microsoft/installer.ts:11]: The URL in my browser shows `macOS`, but I see `macos` works as well (IIS? 🙂)

image

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [inline src/distributions/microsoft/installer.ts:65]:
```suggestion
        `Could not find satisfied version for SemVer ${range}. ${availableOptionsMessage}`
```

I see this error message both ways in the codebase, but let's use the official stylization for new code

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [inline src/distributions/microsoft/installer.ts:76]:
```suggestion
    const jdkVersions = [
```

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [inline src/distributions/microsoft/installer.ts:75]: Do we know what it would take to get the versions dynamically?

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [inline src/distributions/microsoft/installer.ts:81]: We expect `fullVersion` to match up with the other properties, but that may be a source of copy-paste bugs in the future.  I think it's better just to pass `fullVersion` as `Array<number>` (like [Zulu does](https://github.com/actions/setup-java/blob/5f00602cd1b2819185d88dc7a1b1985f598c6705/src/distributions/zulu/models.ts#L8)) and have consumers put it in the appropriate format.

Also, looks like 17.0.0 breaks this pattern; the minor and patch versions don't appear in the download URL: https://docs.microsoft.com/en-us/java/openjdk/older-releases#openjdk-17.  Might be good to add a comment for the next person who works on this

File: .github/workflows/e2e-versions.yml
Comment (brcrista) [inline .github/workflows/e2e-versions.yml:17]: Looks like this is causing the [workflow](https://github.com/actions/setup-java/actions/runs/1492164152) to fail:

```
The workflow is not valid. .github/workflows/e2e-versions.yml (Line: 17, Col: 9): Unrecognized named-value: 'matrix'. Located at position 1 within expression: matrix.distribution != 'microsoft' || matrix.version != '8'
```

File: .github/workflows/e2e-versions.yml
Comment (brcrista) [inline .github/workflows/e2e-versions.yml:173]: Same issue here ...

```
.github/workflows/e2e-versions.yml (Line: 173, Col: 9): Unrecognized named-value: 'matrix'. Located at position 2 within expression: (matrix.distribution == 'zulu' && matrix.architecture == 'x86' && matrix.os != 'macos-latest' ) || (matrix.distribution == 'microsoft' && matrix.architecture == 'arm64')
```

File: .github/workflows/e2e-versions.yml
Comment (brcrista) [reply .github/workflows/e2e-versions.yml:17]: Seems like `matrix` must not be usable in `if`?  Let me double check

**Review by brcrista — COMMENTED** [review]
Looks good!  [This workflow is failing](https://github.com/actions/setup-java/actions/runs/1492164152); I'm looking into why that is (and why it doesn't report on the Conversation page in the PR)

File: .github/workflows/e2e-versions.yml
Comment (brcrista) [reply .github/workflows/e2e-versions.yml:17]: Yep, this is why.  `if` at the job level is evaluated before the matrix is expanded, so it can't reference the `matrix` context.  `name` is evaluated when the job runs, so that's why you can use it there.

Instead, you can use `exclude` to knock a leg out of the matrix: https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#example-excluding-configurations-from-a-matrix

File: README.md
Comment (brendandburns) [reply README.md]: done.

File: src/distributions/microsoft/installer.ts
Comment (brendandburns) [reply src/distributions/microsoft/installer.ts:11]: I don't think there's anything needed here. (Aren't all URLs case-insensitive?)

File: src/distributions/microsoft/installer.ts
Comment (brendandburns) [reply src/distributions/microsoft/installer.ts:75]: Unfortunately there is no easy way today other than scraping HTML which is very fragile (and annoying to implement)

I'd rather have a fixed setting for now.

Comment (brcrista) [reply src/distributions/microsoft/installer.ts:75]: Yep, fixed setting is fine to get started.  Was just curious if we had a lead on next steps

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [reply src/distributions/microsoft/installer.ts:11]: I believe the path part of a URL is usually case-sensitive: https://webmasters.stackexchange.com/questions/90339/why-are-urls-case-sensitive

It seems to work though, so no not a big deal.

File: src/distributions/microsoft/installer.ts
Comment (brendandburns) [reply src/distributions/microsoft/installer.ts]: done.

File: .github/workflows/e2e-versions.yml
Comment (brendandburns) [reply .github/workflows/e2e-versions.yml:183]: done.

File: .github/workflows/e2e-versions.yml
Comment (brendandburns) [reply .github/workflows/e2e-versions.yml]: done.

**Comment by brendandburns** [comment]
@brcrista thanks for the review, I think I have addressed all comments.

**Comment by brcrista** [comment]
Thanks @brendanburns , just kicked off the CI

**Comment by brcrista** [comment]
Looks like there are some unresolved merge conflicts in `.github/workflows/e2e-versions.yml` and linting is failing.  I can push fixes for those issues, if you don't mind.

**Comment by brcrista** [comment]
https://github.com/brendandburns/setup-java/pull/1 should fix the CI runs.  Once those are green we can merge :ship:

**Comment by brendandburns** [comment]
@brcrista Thanks! I merged your PR.

File: src/distributions/microsoft/installer.ts
Comment (brunoborges) [reply src/distributions/microsoft/installer.ts:77]: We plan to publish an API, but it is unclear we will get that done before the January PSU. A hardcoded version list, for now, is fine. Besides, Java updates happen 4 times a year, once every quarter. It is not that bad :-)

File: .github/workflows/e2e-versions.yml
Comment (brunoborges) [inline .github/workflows/e2e-versions.yml:24]: Can we replace 16 with 17 here, since 17 is the latest LTS now?

File: .github/workflows/e2e-versions.yml
Comment (brunoborges) [inline .github/workflows/e2e-versions.yml:189]: Should we move up to 17?

File: README.md
Comment (brunoborges) [inline README.md:62]: Correct name is `Microsoft Build of OpenJDK`

File: .github/workflows/e2e-versions.yml
Comment (brunoborges) [inline .github/workflows/e2e-versions.yml:25]: I think we should also exclude Microsoft's version 16, as it is not an LTS release and we will no longer provide updates.

File: __tests__/distributors/microsoft-installer.test.ts
Comment (brunoborges) [inline __tests__/distributors/microsoft-installer.test.ts:22]: Given 16 is not an LTS release of MS Build of OpenJDK, I think we should not list it here.

File: src/distributions/microsoft/installer.ts
Comment (brunoborges) [inline src/distributions/microsoft/installer.ts:86]: See previous comments regarding making 16 available.

File: src/distributions/microsoft/installer.ts
Comment (brcrista) [reply src/distributions/microsoft/installer.ts:77]: Ok awesome, having an API will be good.  The main functionality this unblocks is being able to install any version that's been published from the whole history.  Regarding updates, people expect to be able to use the newest versions of tools with Actions as soon as they come out, so that also makes sure we're always up-to-date.

Not sure what your plans are for an API, but other tools (such as Node) just host a JSON file with version info and links to the downloads.  [Example](https://nodejs.org/dist/index.json)

File: .github/workflows/e2e-versions.yml
Comment (brendandburns) [reply .github/workflows/e2e-versions.yml:24]: Probably worth doing in a different PR since it is unrelated to adding MSFT.

File: .github/workflows/e2e-versions.yml
Comment (brendandburns) [reply .github/workflows/e2e-versions.yml:198]: As above.

File: README.md
Comment (brendandburns) [reply README.md]: Done (apologies, this got lost in my rebasing)

File: .github/workflows/e2e-versions.yml
Comment (brendandburns) [reply .github/workflows/e2e-versions.yml:25]: I think for now, since Microsoft hosts it, we should keep it in the spirit of meeting users where they are. Many users haven't updated their CI/CD for JDK 17.

File: __tests__/distributors/microsoft-installer.test.ts
Comment (brendandburns) [reply __tests__/distributors/microsoft-installer.test.ts:22]: as above, but if you feel strongly I will remove it.

File: src/distributions/microsoft/installer.ts
Comment (brendandburns) [reply src/distributions/microsoft/installer.ts:84]: as above.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

File: .github/workflows/e2e-versions.yml
Comment (brunoborges) [reply .github/workflows/e2e-versions.yml:24]: Ah, yes. That'd be better indeed.

File: .github/workflows/e2e-versions.yml
Comment (brunoborges) [reply .github/workflows/e2e-versions.yml:25]: Yeah, that makes sense. We are not exposing support timelines for any of the distros through the GH Action anyways, so, it is up to the end-user to check the documentation of each distro.

File: .github/workflows/e2e-versions.yml
Comment (brunoborges) [reply .github/workflows/e2e-versions.yml:198]: Noted.

File: __tests__/distributors/microsoft-installer.test.ts
Comment (brunoborges) [reply __tests__/distributors/microsoft-installer.test.ts:22]: No, it's fine. See previous comment.

File: __tests__/distributors/microsoft-installer.test.ts
Comment (dmitry-shibanov) [inline __tests__/distributors/microsoft-installer.test.ts:35]: Could you please change `var` to `let`.

File: src/distributions/microsoft/installer.ts
Comment (dmitry-shibanov) [inline src/distributions/microsoft/installer.ts:13]: The supportedArchitecture variable is not used. Could you please remove it.

File: src/distributions/microsoft/installer.ts
Comment (dmitry-shibanov) [inline src/distributions/microsoft/installer.ts:73]: Could you please add console.timeEnd or remove it for now.

File: src/distributions/microsoft/models.ts
Comment (dmitry-shibanov) [inline src/distributions/microsoft/models.ts:2]: Could you please remove the export keyword because these types are only used inside `models.ts`

**Comment by dmitry-shibanov** [comment]
Could you please revert the changes for package-lock.json and use these commands `npm ci` -> `npm run build`.

**Comment by brcrista** [comment]
Looking at my fork, looks like there will be a couple more issues:

https://github.com/brcrista/setup-java/runs/4408213781?check_suite_focus=true#step:4:7

```
Found java version: __tests__/verify-java.sh: line 24: /c/hostedtoolcache/windows/Java_Microsoft_jdk/11.0.13/aarch64/bin/java: cannot execute binary file: Exec format error
Error: Unexpected version
```

Similar issue for Ubuntu as well.

@brendanburns if you want to [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) on this PR then we can push fixes for these issues

image

File: __tests__/distributors/microsoft-installer.test.ts
Comment (brendandburns) [reply __tests__/distributors/microsoft-installer.test.ts]: done.

File: src/distributions/microsoft/installer.ts
Comment (brendandburns) [reply src/distributions/microsoft/installer.ts]: done.

File: src/distributions/microsoft/installer.ts
Comment (brendandburns) [reply src/distributions/microsoft/installer.ts]: removed.

**Comment by brendandburns** [comment]
@dmitry-shibanov comments addressed.

@brcrista not sure where the aarch is coming from but it seems that actions doesn't support that architecture? I have ticked the "allow edits from maintainers" box, so any fixes or pointers would be great.

Thanks!

File: .github/workflows/e2e-versions.yml
Comment (dmitry-shibanov) [inline .github/workflows/e2e-versions.yml:197]: Could you please resolve merge conflicts.

**Review by brcrista — APPROVED** [review]
Looks like the `Licensed` workflow is failing, but that's not related to these changes https://github.com/actions/setup-java/actions/runs/1555484816

Thank you @brendanburns and @brunoborges, this is a huge help!

**Comment by brendandburns** [comment]
@brcrista thank you for your help getting this across the finish line!!

**Comment by ethomson** [comment]
Thanks so much @brendanburns!

**Comment by sullis** [comment]
I just tried using the Microsoft JDK in a GitHub Actions workflow. It did not work.  My workflow uses 'setup-java@v2' and there hasn't been a release that incorporates the new Microsoft JDK code.

Can somebody tag a new release of setup-java?  

cc: @ethomson @brendandburns

**Comment by brunoborges** [comment]
@sullis you can reference the commit hash as a tag

**Pull request merged**
